src/midi/midi_mt32.cpp

```diff
@@ -683,101 +694,124 @@ void MidiHandler_mt32::Close()
+	last_rendered_ms += (frames * ms_per_frame);
+	return check_cast<uint16_t>(frames);
 }
 
+// The request to play the channel message is placed in the MIDI work FIFO
 void MidiHandler_mt32::PlayMsg(const uint8_t *msg)
 {
-	assert(channel);
-	channel->WakeUp();
-
-	const auto msg_words = reinterpret_cast<const uint32_t *>(msg);
-	const std::lock_guard<std::mutex> lock(service_mutex);
-	service->playMsgAt(SDL_SwapLE32(*msg_words), GetMidiEventTimestamp());
+	constexpr auto channel_len = 4;
+	std::vector<uint8_t> message(msg, msg + channel_len);
+	MidiWork work{std::move(message), GetNumPendingFrames(), MessageType::Channel};
+	work_fifo.Enqueue(std::move(work));
 }
 
+// The request to play the sysex message is placed in the MIDI work FIFO
 void MidiHandler_mt32::PlaySysex(uint8_t *sysex, size_t len)
 {
-	assert(channel);
-	channel->WakeUp();
-
-	assert(len <= UINT32_MAX);
-	const auto msg_len = static_cast<uint32_t>(len);
-	const std::lock_guard<std::mutex> lock(service_mutex);
-	service->playSysexAt(sysex, msg_len, GetMidiEventTimestamp());
+	std::vector<uint8_t> message(sysex, sysex + len);
+	MidiWork work{std::move(message), GetNumPendingFrames(), MessageType::Sysex};
+	work_fifo.Enqueue(std::move(work));
 }
 
 // The callback operates at the frame-level, steadily adding samples to the
 // mixer until the requested numbers of frames is met.
-void MidiHandler_mt32::MixerCallBack(uint16_t requested_frames)
+void MidiHandler_mt32::MixerCallBack(const uint16_t requested_frames)
 {
-	while (requested_frames) {
-		const auto frames_to_be_played = std::min(GetRemainingFrames(),
-		                                          requested_frames);
-		const auto sample_offset_in_buffer = play_buffer.data() +
-		                                     last_played_frame * 2;
-		channel->AddSamples_sfloat(frames_to_be_played, sample_offset_in_buffer);
-		requested_frames -= frames_to_be_played;
-		last_played_frame += frames_to_be_played;
+	assert(channel);
+
+	constexpr auto warning_percent = 5;
+	if (const auto percent_full = frame_fifo.GetPercentFull(); percent_full < warning_percent) {
+		static auto iteration = 0;
+		if (iteration++ % 100 == 0) {
+			LOG_WARNING("MT32: Audio FIFO is %2d%% full", percent_full);
+		}
 	}
+
+	static std::vector<AudioFrame> frames = {};
+	frame_fifo.BulkDequeue(frames, requested_frames);
+	channel->AddSamples_sfloat(requested_frames, &frames[0][0]);
+
+	last_rendered_ms = PIC_FullIndex();
 }
 
-// Returns the number of frames left to play in the buffer.
-uint16_t MidiHandler_mt32::GetRemainingFrames()
+void MidiHandler_mt32::RenderFramesToFifo(const uint16_t num_frames)
 {
-	// If the current buffer has some frames left, then return those ...
-	if (last_played_frame < FRAMES_PER_BUFFER)
-		return FRAMES_PER_BUFFER - last_played_frame;
+	static std::vector<AudioFrame> frames = {};
+	// Maybe expand the vector
+	if (frames.size() < num_frames) {
+		frames.resize(num_frames);
+	}
```

Comment:
Ok, it's obvious at this point that frames mean sample frames. I'd still recommend to clarify it in this file to avoid initial confusion.